config-system.lisp

```diff
@@ -0,0 +1,157 @@
+        (format s  "Setting ~A not found." (%full-symbol-string place-symbol)))))))
+
+(defun %generate-config-var-code (declare-type name default validator documentation)
+  (check-type documentation string)
+  (with-gensyms (default-value is-valid func)
+    `(progn
+       (,declare-type ,name ,default ,documentation)
+       (let* ((,default-value ,name)
+              (,func ,validator)
+              (,is-valid (funcall ,func ,default-value)))
+         (if ,is-valid
+             (setf (gethash (quote ,name) *config-vars*)
+                   (make-config-info :name (quote ,name) :default ,default-value
+                                     :validator ,func :doc ,documentation))
+             (error 'invalid-datum-error :place (quote ,name) :value ,default-value))))))
+
+(defmacro define-config-parameter (name default validator documentation)
+  "Create and register a configurable variable with the given default value,
+validator function, and documentation using DEFPARAMETER"
+  ;; if we are given a lambda, we need to evaluate it in order to call it
+  (%generate-config-var-code 'defparameter name default validator documentation))
+
+(defmacro define-config-var (name default validator documentation)
+  "Create and register a configurable variable with the given default value,
+validator function, and documentation using DEFVAR"
+  ;; if we are given a lambda, we need to evaluate it in order to call it
+  (%generate-config-var-code 'defvar name default validator documentation))
+
+(defmacro define-config-enum (name default values documentation)
+  "Same as DEFINE-CONFIG-PARAMETER, except the validation function
+is built by creating a function that checks if the set  value is in the VALUES list."
+  ;; if we wanted to get fancy, we could define a special type for this enumeration at this point too.
+  `(define-config-parameter ,name ,default (lambda (x) (member x ,values :test #'equal))
+                            ,documentation))
+
+(defun list-all-configurations ()
+  "List all of the available customizable settings"
+  (hash-table-values *config-vars*))
+
+(defun %find-possible-settings (setting-name table)
+  "Find the settings that have the same symbol name of SETTING-NAME but are in a different package"
+  (declare (type hash-table table)
+           (type symbol setting-name))
+  (let ((name (symbol-name setting-name)))
+    ;; for some reason, the package isn't always included with the symbol name even
+    ;; if the top level isn't in the symbol's package, so we need to manually get the symbol name.
+    (mapcar #'%full-symbol-string
+            (remove-if-not (lambda (x) (string-equal (symbol-name x) name))
+                           (mapcar #'config-info-name (hash-table-values table))))))
+
+(defun get-configuration-info (setting-name)
+  "Find the info for the config variable stored in the symbol SETTING-NAME."
+  (declare (type symbol setting-name))
+  (gethash setting-name *config-vars*))
```

Comment:
The functions have the hash table hard-coded in, which will probably lead to problems in the future.

Reply:
What you are talking about is something that [`destructring-bind`](http://clhs.lisp.se/Body/m_destru.htm) provides. In this case though, I think an optional parameter with the default of `*config-vars*` would be fine.